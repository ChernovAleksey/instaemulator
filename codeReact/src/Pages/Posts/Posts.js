import PostCard from "../../Components/PostCard/PostCard";
import Following from "../../Components/Following/Following";
import { useSelector } from "react-redux";
import { useState } from "react";
import InfiniteScroll from "react-infinite-scroll-component";
import "./Posts.scss";

function Posts() {
  const [numb, setNumb] = useState(2);
  const posts = useSelector((store) => store.reducer.postsarray);
  const demo = posts.slice(0, numb);
  const generator = () => {
    setNumb((prev) => prev + 2);
  };

  return (
    <div className="post_maincontainer">
      <InfiniteScroll dataLength={demo.length} next={generator} hasMore={true}>
        {demo.map(
          ({ id, likes, url, userIndex, hasBackground, _id, comments }) => (
            <PostCard
              key={id}
              id={id}
              comments={comments}
              hasBackground={hasBackground}
              likes={likes}
              url={url}
              userIndex={userIndex}
              _id={_id}
            />
          )
        )}
      </InfiniteScroll>

      <Following />
    </div>
  );
}
export default Posts;
