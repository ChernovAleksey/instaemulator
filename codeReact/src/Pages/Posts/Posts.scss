.post_maincontainer{
    padding: 120px 25%;
    min-height: 600px;

}