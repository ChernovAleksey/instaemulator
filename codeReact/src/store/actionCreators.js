import {
  GET_POSTS_ARRAY,
  GET_USERS_ARRAY,
  USER_IS_FOLLOWING,
  USER_IS_UNFOLLOWING,
  GIVING_LIKE,
  TAKING_LIKE,
  ADDING_COMMENT,
  IS_MODAL_OPEN,
  CHOSEN_TO_OPERATE,
} from "./actions";

export const getPostsArrayAC = () => async (dispatch) => {
  let postsarray = [];
  const { data } = await fetch(process.env.REACT_APP_API_URL_POSTS).then(
    (res) => res.json()
  );
  postsarray = data;
  dispatch({ type: GET_POSTS_ARRAY, payload: postsarray });
};

export const getUsersArrayAC = () => async (dispatch) => {
  let usersarray = [];
  const { data } = await fetch(process.env.REACT_APP_API_URL_USERS).then(
    (res) => res.json()
  );
  usersarray = data;
  dispatch({ type: GET_USERS_ARRAY, payload: usersarray });
};

export const isUserFollowingAC =
  ({ _id }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_USERS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { isFollower: true } }),
    });
    dispatch({ type: USER_IS_FOLLOWING, payload: _id });
  };

export const isUserUnFollowingAC =
  ({ _id }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_USERS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { isFollower: false } }),
    });
    dispatch({ type: USER_IS_UNFOLLOWING, payload: _id });
  };

export const givingLikeAC =
  ({ _id, likes }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        id: _id,
        target: { hasBackground: true, likes: likes + 1 },
      }),
    });
    dispatch({ type: GIVING_LIKE, payload: _id });
  };

export const takingLikeAC =
  ({ _id, likes }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        id: _id,
        target: { hasBackground: false, likes: likes - 1 },
      }),
    });
    dispatch({ type: TAKING_LIKE, payload: _id });
  };

export const addingCommentAC =
  ({ _id, array }) =>
  async (dispatch) => {
    fetch(process.env.REACT_APP_API_URL_POSTS, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id: _id, target: { comments: array } }),
    });
    dispatch({ type: ADDING_COMMENT, payload: { _id, array } });
  };

export const isModalOpenAC = (payload) => ({ type: IS_MODAL_OPEN, payload });

export const chosenToOperateAC = (payload) => ({
  type: CHOSEN_TO_OPERATE,
  payload,
});
