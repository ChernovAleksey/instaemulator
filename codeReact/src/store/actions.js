export const GET_POSTS_ARRAY = "GET_POSTS_ARRAY";
export const GET_USERS_ARRAY = "GET_USERS_ARRAY";
export const USER_IS_FOLLOWING = "USER_IS_FOLLOWING";
export const USER_IS_UNFOLLOWING = "USER_IS_UNFOLLOWING";
export const GIVING_LIKE = "GIVING_LIKE";
export const TAKING_LIKE = "TAKING_LIKE";
export const ADDING_COMMENT = "ADDING_COMMENT";
export const IS_MODAL_OPEN = "IS_MODAL_OPEN";
export const CHOSEN_TO_OPERATE = "CHOSEN_TO_OPERATE";

