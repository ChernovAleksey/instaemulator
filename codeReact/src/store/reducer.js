import {
  GET_POSTS_ARRAY,
  GET_USERS_ARRAY,
  USER_IS_FOLLOWING,
  USER_IS_UNFOLLOWING,
  GIVING_LIKE,
  TAKING_LIKE,
  ADDING_COMMENT,
  CHOSEN_TO_OPERATE,
  IS_MODAL_OPEN,
} from "./actions";

const initialValue = {
  usersarray: [],
  postsarray: [],
  isModalOpen: false,
  chosenToOperate: {},
};
const reducer = (state = initialValue, action) => {
  switch (action.type) {

    case GET_POSTS_ARRAY: {
      return { ...state, postsarray: action.payload };
    }
      
    case GET_USERS_ARRAY: {
      return { ...state, usersarray: action.payload };
    }
      
    case USER_IS_FOLLOWING: {
      const newArray = [...state.usersarray];
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].isFollower = true;
      }
      return { ...state, array: newArray };
    }
      
    case USER_IS_UNFOLLOWING: {
      const newArray = [...state.usersarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].isFollower = false;
      }
      return { ...state, array: newArray };
    }
      
    case GIVING_LIKE: {
      const newArray = [...state.postsarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].hasBackground = true;
        newArray[index].likes = newArray[index].likes + 1;
      }
      return { ...state, array: newArray };
    }
      
    case TAKING_LIKE: {
      const newArray = [...state.postsarray];
      console.log(newArray);
      const index = newArray.findIndex((el) => el._id === action.payload);
      if (index !== -1) {
        newArray[index].hasBackground = false;
        newArray[index].likes = newArray[index].likes - 1;
      }
      return { ...state, array: newArray };
    }
      
    case ADDING_COMMENT: {
      const newArray = [...state.postsarray];
      const index = newArray.findIndex((el) => el._id === action.payload._id);
      if (index !== -1) {
        newArray[index].comments = action.payload.array;
      }
      return { ...state, array: newArray };
    }
      
    case IS_MODAL_OPEN: {
      return { ...state, isModalOpen: action.payload };
    }
      
    case CHOSEN_TO_OPERATE: {
      return { ...state, chosenToOperate: action.payload };
    }

    default: {
      return state;
    }
  }
};

export default reducer;
