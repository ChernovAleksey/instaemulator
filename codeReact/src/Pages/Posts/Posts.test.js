import Posts from './Posts';
import { render} from "@testing-library/react";
import { Provider} from "react-redux";
import store from "../../store/index";

const Component = () => {
  return (
    <>
      <Posts />
    </>
  );
};
const RenderComponent = () => {
  return (
    <Provider store={store}>
      <Posts />
    </Provider>
  );
};
describe("Posts renders", () => {
  test("should Posts match snapshot", () => {
    const { asFragment } = render(<RenderComponent />);
    expect(asFragment()).toMatchSnapshot();
  });
});