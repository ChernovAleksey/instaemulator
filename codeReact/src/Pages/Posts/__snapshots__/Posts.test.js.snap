// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts renders should Posts match snapshot 1`] = `
<DocumentFragment>
  <div
    class="post_maincontainer"
  >
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      />
    </div>
    <div
      class="following_maincontainer"
    >
      <div
        class="following-stories"
      >
        <p>
          stories for you
        </p>
      </div>
      <div
        class="following-recomend"
      >
        <p>
          recomended for you
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
